src/GraphNavigator/DeserializationGraphNavigator.php

```diff
@@ -212,12 +214,21 @@ public function accept($data, ?array $type = null)
                     $this->context->pushPropertyMetadata($propertyMetadata);
                     try {
                         $v = $this->visitor->visitProperty($propertyMetadata, $data);
+
                         $this->accessor->setValue($object, $v, $propertyMetadata, $this->context);
                     } catch (NotAcceptableException $e) {
                         if (true === $propertyMetadata->hasDefault) {
                             $cloned = clone $propertyMetadata;
                             $cloned->setter = null;
                             $this->accessor->setValue($object, $cloned->defaultValue, $cloned, $this->context);
+                        } elseif (!$allowsNull && $this->visitor->getRequireAllRequiredProperties()) {
```

Comment:
I'm not sure if we should handle this here... could you please explain why do you need this check at this point?

Reply:
> I'm not sure if we should handle this here... could you please explain why do you need this check at this point?

I added this check to make the normal deserialization process is less permissive, resulting in more accurate type detection.

Specifically, if the object you are deserializing into contains a required field, and the JSON does not contain that field, the existing deserializer will deserialize without complaint.  While, that behaviour could be desirable in some cases, and presumably must be maintained for backwards compatibility, it's undesirable for deserializing non-discriminated unions.

For non-discriminated unions, we use a "guess and check" approach to deserialization, and the absence of a required field is a strong signal that the data we're deserializing may better match a different type.


```
class SimplePerson {
    public string $name;
    public ?string $nickname;
}

class SimpleCar {
    public string $model;
    public ?string $nickname;
}

class WrapperClass {
  public SimplePerson|SimpleCar $entity
}  
```

If we want to deserialize the following data:

```
{
  "entity": {
    "name": "ian"
  }
}
```

Without such a rule, the resulting class is determined by the order of application.  If you consider `SimplePerson` first, then all is well.  If you consider `SimpleCar` first, then the deserialization model skips `name` because it doesn't exist in the class, and deserialization will succeed returning a null `SimpleCar`.

By adding the rule,  when `SimpleCar` is tested, the required `model` attribute is not present, so deserialization will fail.

I agree that the implementation is awkward, and would be happy to refactor!

Reply:
I would really avoid implementing such usecase. It is too specific and I'm really not sure if it worth the complexity.

```json
{
  "entity": {
    "type": "simple_person",
    "name": "ian"
  }
}```

A `type` filed should be always present in order to match efficiently the class name to use.

The same type of discussion were taken when implementing the `@DiscrminatorMap` for inheritance and even in that case any other solution was just not worth the effort.

Reply:
I think it's worth the effort, and I don't think it's a very specific usecase.

While, of course deserializing discriminated unions is easier, it's not the only type of data that exists.  It's really easy to say "just change the data", but in many cases, that isn't possible - for example: if someone's deserializing data from an external API.  Anecdotally in my situation, this case comes up regularly.

In order to provide good support for PHP 8, jms/serializer should have good support for union types,.

I'll note that other deserialization libraries do provide this functionality: `dataclasses_json`, `pydantic`, `zod` are just three examples.

I've provided an implementation that is opt-out (by not using the `UnionHandler` altogether, or you can construct the `UnionHandler` with `requireAllProperties=false` to avoid the new failure logic).  Is the concern the added code complexity?  I'm happy to refactor the code.

However, if you still think this isn't an appropriate feature fro the jms/serializer - I would ask that at least we provide a workaround for users.  If we could include the ability to `requireAllRequiredFields` when deserializing, it would allow me to write a custom UnionHandler outside of jms/serializer.